### ISSUE TITLE
Karpenter v1.0.0 chart does not respect Helm values in packaged CRDs leading to ArgoCD upgrade difficulty

### ISSUE THREAD

**Issue by booleanbetrayal** [aws/karpenter-provider-aws#6847]
### Description

**Observed Behavior**:

The setting of any `webhook` values (such as `webhook.serviceName`) in the `karpenter` primary chart is not respected. Instead, the chart embeds static CRDs, which are [symlinked in this repository](https://github.com/aws/karpenter-provider-aws/tree/main/pkg/apis/crds), unlike the `karpenter-crd` chart, [which respects these values](https://github.com/aws/karpenter-provider-aws/tree/main/charts/karpenter-crd/templates). 

This makes upgrading from v0.37.0 to v1.0.0 using platforms like ArgoCD difficult and error-prone since ArgoCD is capable of managing CRDs and will likewise detect drift if they are updated manually or via a secondary chart such as `karpenter-crd`.

In our case, the `conversion.webhook.clientConfig.service.name` value for the conversion webhook (specified in the static CRDs) deployed with the `karpenter` chart did not match the name that the chart generates for the Service itself (using the `karpenter.fullName` [helper](https://github.com/aws/karpenter-provider-aws/blob/main/charts/karpenter/templates/service.yaml#L4)), and so all webhook calls would fail. Likewise, disabling the webhook could not be done with values and manual edits will show diff changes in ArgoCD.

Attempting to utilize both charts in ArgoCD will attempt to create duplicate resources and could cause unexpected issues. There is no way to disable the deployment of CRDs in the `karpenter` primary chart. 

A possible workaround for our specific issue would be to set the `nameOverride` value to `karpenter` such that it matches the static `serviceName` buried in the CRD, but ideally we don't even need these conversion webhooks and would prefer to have them disabled entirely.

**Expected Behavior**:

The `karpenter` Helm package respects Helm values in its CRDs, just as the `karpenter-crd` Chart does. If static CRDs are required they can be referenced elsewhere from the GitHub repository, etc.

Alternatively (and less elegantly), a dual-chart solution with a flag supporting the disabling of CRDs in the primary chart (or just not packaging them to begin with) could also be a possible fix vector, as it would allow the other chart to install without duplicates.

**Reproduction Steps** (Please include YAML):

Attempt to upgrade the `karpenter` chart from `v0.37.0` to `v1.0.0` using ArgoCD. Specify override Helm values for `webhook` properties and observe no diff is presented in the CRD definitions reflecting those overrides. 

**Versions**:
- Chart Version: 1.0.0
- Kubernetes Version (`kubectl version`): 1.30.2

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

**Comment by engedaam** [comment]
As part of ArgoCD, you should be able to disable CRD installation, no? https://argo-cd.readthedocs.io/en/stable/user-guide/helm/#helm-skip-crds. Also, It sounds like you updated the `fullnameOverride`? Does that align with what is passed as the webhook.serviceName in `karpenter-crd` charts?

**Comment by booleanbetrayal** [comment]
Hi @engedaam and thanks for the quick feedback. That `skip-crds` setting applies to an ArgoCD Application, but that might deploy multiple charts / sub-charts. In our case, we have other Helm packages being deployed in this namespace / ArgoCD Application (`kube-system`) which are modeled in this fashion. We wouldn't want them to all carry this setting, as this is the only problematic deployment.

Other popular charts often include some flag to toggle CRD deployment for more granular fidelity. For example, [ESO uses a `installCRDs` flag for this feature](https://github.com/external-secrets/external-secrets/blob/main/deploy/charts/external-secrets/values.yaml#L34C1-L35C18).

We should be able to update the `fullnameOverride` for the `karpenter` chart to match the `"karpenter"` value for `conversion.webhook.clientConfig.service.name` that its bundled (static) CRDs set, and hopefully everything just works at that point since the Service would be named appropriately. However, that requires rebuilding all our existing Karpenter resources to support the new naming, along with IAM trust policies, etc.

At this point, I'm curious how people running Karpenter within an ArgoCD Application are performing the v1.0.0 upgrade.

**Comment by booleanbetrayal** [comment]
> At this point, I'm curious how people running Karpenter within an ArgoCD Application are performing the v1.0.0 upgrade.

Just to provide more context, [Helm won't even overwrite resources that are managed-by ArgoCD](https://github.com/argoproj/argo-cd/issues/1672), and I'm not sure how to spoof [Helm's `.Release.Service` stamping](https://helm.sh/docs/chart_best_practices/labels/). As a result, AFAIK, the official [v1 Migration Guide](https://karpenter.sh/docs/upgrading/v1-migration/) can't be followed for ArgoCD Karpenter deployments and the CRD upgrade step. Even if you could spoof it so Helm CLI did deploy those `karpenter-crd` CRDs, the Karpenter CRDs would immediately show up as drifting out of sync with the `karpenter` Chart source in ArgoCD Application display, as they would differ from the bundled variants.

**Comment by Vinaum8** [comment]
+1

**Comment by Vinaum8** [comment]
The expected behavior according to this issue would be perfect, as we would have the option of enabling or not the installation of CRDs with the karpenter helm chart.

**Comment by artem-nefedov** [comment]
Given 75d482f6004540fabd8c9b2eb2cc602742b96ca9 is already in main, I assume official guidelines from now on will be "disable CRDs in the main chart and always use karpenter-crds if you require any sort of CRD customization".

**Comment by booleanbetrayal** [comment]
@artem-nefedov - How does one disable CRDs in the main chart? `main` is still currently bundling all the static CRDs as it was before, without any mechanism to prevent deployment in ArgoCD, as far as I can tell.

**Comment by artem-nefedov** [comment]
> @artem-nefedov - How does one disable CRDs in the main chart? `main` is still currently bundling all the static CRDs as it was before, without any mechanism to prevent deployment in ArgoCD, as far as I can tell.

1. With helm CLI, you would use `--skip-crds` option.
2. With Flux2 GitOps, you would set `spec.install.crds: Skip` and `spec.upgrade.crds: Skip` in HelmRelease.
3. I don't use ArgoCD, but quick search found that it supports feature to skip CRDs since January 2022: https://github.com/argoproj/argo-cd/pull/8012

**Comment by booleanbetrayal** [comment]
@artem-nefedov - I've posted a response to that workaround [elsewhere](https://github.com/aws/karpenter-provider-aws/issues/6765#issuecomment-2307072556) but the gist is that there may be multiple Helm charts comprising an ArgoCD Application. Skipping CRDs to relieve a problematic deployment from one Chart could cascade into multiple issues with the maintenance of CRDs in other charts.

**Comment by artem-nefedov** [comment]
@booleanbetrayal  Maybe I'm missing something, but so far this looks like solvable problem. Split it into two separate Applications, and optionally add a third top-level Application that installs both if you still want to have a single entrypoint.

**Comment by booleanbetrayal** [comment]
> @booleanbetrayal Maybe I'm missing something, but so far this looks like solvable problem. Split it into two separate Applications, and optionally add a third top-level Application that installs both if you still want to have a single entrypoint.

@artem-nefedov - So if my Application is `kube-system`, deploying to the `kube-system` Namespace and I am running Karpenter within it, along with multiple other Charts (AWS-LB, ESO, etc) that typically control their CRD deployments with things like [Helm flags](https://github.com/external-secrets/external-secrets/blob/main/deploy/charts/external-secrets/values.yaml#L34C1-L35C18), I am supposed to re-architect my application to move / rename resources into a separate Applications and Namespaces in all my clusters, juggling Karpenter controller uptime so that Nodes can continue to spool, all in order to support this v1.0.0 upgrade? 

You might not have much familiarity with ArgoCD, but I can tell you that this is not a light chore, would be prone to outage chances, is not documented as a migration vector, and would feel especially egregious in light of AWS's recent [endorsement](https://aws.amazon.com/blogs/containers/continuous-deployment-and-gitops-delivery-with-amazon-eks-blueprints-and-argocd/) [of ArgoCD](https://aws.amazon.com/blogs/containers/multi-cluster-management-for-kubernetes-with-cluster-api-and-argo-cd/) [deployment approaches](https://aws.amazon.com/blogs/containers/cloud-native-ci-cd-with-tekton-and-argocd-on-aws/).

**Comment by jmdeal** [comment]
> that typically control their CRD deployments with things like [Helm flags](https://github.com/external-secrets/external-secrets/blob/main/deploy/charts/external-secrets/values.yaml#L34C1-L35C18)

It looks like these charts don't follow [Helm v3's recommendations](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/) for CRD management. These recommendations require that CRDs in the main chart be added to the `crds` directory, which cannot be templated. If templating is required, like it is for Karpenter's conversion webhook configuration, a separate CRD chart should be used. Without good justification, we shouldn't diverge from Helm's standard. 

> a dual-chart solution with a flag supporting the disabling of CRDs in the primary chart

This is what `--skip-crds` is explicitly for. I was surprised to hear this wasn't supported at a per-chart level, but that doesn't appear to be the case to me. `skipCrds` is part of the source spec and I was able to get this working with the following application spec:


Following [Helm v3's recommendations](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/) for CRD management, it is not possible to template a CRD in the main chart. We provide a non-templated CRD in the main chart which works for Karpenter's defaults, but the CRD chart will need to be used if templating is required. As far as I can tell, other solutions which allow installation of templated CRDs in a main chart are working around Helms CRD management design which we are adhering to. `--skip-crds` is the correct method to skip CRDs in the main chart. 

> That skip-crds setting applies to an ArgoCD Application, but that might deploy multiple charts / sub-charts.

This does not appear to be the case, `skipCrds` is a field on the source, not on the application. I was able to get this working with the following application spec:

```yaml
project: default
destination:
  server: https://kubernetes.default.svc
  namespace: karpenter
sources:
  - repoURL: public.ecr.aws/karpenter
    targetRevision: 1.0.1
    chart: karpenter-crd
  - repoURL: public.ecr.aws/karpenter
    targetRevision: 1.0.1
    helm:
      parameters:
        - name: settings.clusterName
          value: karpenter-oss-dev
        - name: serviceAccount.annotations.eks\.amazonaws\.com/role-arn
          value: arn:aws:iam::862780986519:role/karpenter-oss-dev-karpenter
      skipCrds: true
    chart: karpenter
```
While I agree that it would be nicer to handle the installation through a single chart, we don't want to diverge from Helm's recommendations to do so. Is there anything the `skipCrds` solution does not address.

**Comment by jmdeal** [comment]
For context, we've went with this design following [Helm v3's recommendations](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/) for managing CRDs. These recommendations state that, in a main chart, CRDs should be placed in the `crds` directory. Resources in this directory can not be templated, which is why the packaged CRDs are hardcoded with the package defaults (this only affects the conversion block). When templating is required, the recommendation is to use a separate chart, in this case `karpenter-crd`. 

> Other popular charts often include some flag to toggle CRD deployment for more granular fidelity.

The linked examples in this thread appear to be working around Helm's design for managing CRDs. Unless absolutely necessary, we shouldn't go down this path and stick to helm standard practices. 

>  Skipping CRDs at an application level is not an option.

This does not appear to be necessary. `skipCrds` is scoped to the source, not the application. You should be able to skip CRDs in one source and not another within a single application. I was able to verify with the following application; CRDs were not installed from the main Karpenter chart, but were installed from the aws-controllers-k8s chart:

```yaml
project: default
destination:
  server: https://kubernetes.default.svc
  namespace: karpenter
sources:
  - repoURL: public.ecr.aws/aws-controllers-k8s
    targetRevision: 1.2.17
    chart: ec2-chart
  - repoURL: public.ecr.aws/karpenter
    targetRevision: 1.0.1
    chart: karpenter-crd
  - repoURL: public.ecr.aws/karpenter
    targetRevision: 1.0.1
    helm:
      parameters:
        - name: settings.clusterName
          value: redacted
        - name: serviceAccount.annotations.eks\.amazonaws\.com/role-arn
          value: redacted
      skipCrds: true
    chart: karpenter

```

**Comment by booleanbetrayal** [comment]
@jmdeal - This is a bit of an annoying workaround for users who already have `karpenter` being deployed in an ArgoCD application, originating from a private GitOps repo, containing relevant chart binaries and YAML values. For example, here is the core of our Application definition:

```yaml
project: kube-system
source:
  helm:
    valueFiles:
      - staging-values.yaml
  path: kube-system
  repoURL: user@privately-hosted-git-repo.com/some-org/some-private-gitops-repo
  targetRevision: main
```

In this instance we would either have to expose our Application to third-party supply chain repositories (a deal-breaker for lots of orgs who attempt to mitigate supply-side attacks / reduce third-party dependency failures) or figure out a new scheme of pushing `karpenter` chart + value definitions to a private repo under a new repo, tag, or path.

Even if that were feasible for most users, they would then be forced to perform a moderately risky Application migration to move a pre-1.0.0 running Karpenter controller + provisioning (NodePool, NodeClass) manifests, into the appropriate new source state. Feels like a lot of users could accidentally induce some sort of Karpenter downtime in such a requirement if they fail to atomically swap resources with identical configuration.

If you all absolutely do not wish to implement a convenience flag for this, are adamant on keeping with Helm 3 recommendations, why not just drop the static CRDs from the primary chart and make the `karpenter-crd` chart an explicit or implicit dependency?  **That would also align with the same [Helm 3 Recommendations (option 2)](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/)** which states:

> Another way to do this is to put the CRD definition in one chart, and then put any resources that use that CRD in another chart.
>
> In this method, each chart must be installed separately. However, this workflow may be more useful for cluster operators who have admin access to a cluster

Or alternatively, just drop the conversion webhook altogether and clarify upgrade requirements to appropriate provisioning manifests?

**Comment by thomasgouveia** [comment]
> If you all absolutely do not wish to implement a convenience flag for this, are adamant on keeping with Helm 3 recommendations, why not just drop the static CRDs from the primary chart and make the karpenter-crd chart an explicit or implicit dependency? That would also align with the same [Helm 3 Recommendations (option 2)](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/)

+1. I also mentioned adding a dependency to the `karpenter` chart [here](https://github.com/aws/karpenter-provider-aws/issues/6765#issuecomment-2296238086). Do you have any feedback about it?

**Comment by trojaond** [comment]
@booleanbetrayal  Have you considered including `karpenter-crd` as a separate helm chart in your private repo and reference it from a single ArgoCD application. In this way it does not affect the supply chain.
```
  project: {project}
  destination:
    namespace: kube-system
    server: 'https://kubernetes.default.svc'
  sources:
  - repoURL: {private_git_repository}
    targetRevision: HEAD
    path: '{path_to_helmcharts}/karpenter-crd/'
  - repoURL: {private_git_repository}
    targetRevision: HEAD
    path: '{path_to_helmcharts}/karpenter/'
    helm:
      skipCrds: true
      values: |-
          ...
```
This respect the [Helm 3 Recommendations (option 2)] (https://helm.sh/docs/chart_best_practices/custom_resource_definitions/) as there are 2 charts but they are being deployed from a single ArgoCD Application.

**Comment by booleanbetrayal** [comment]
@trojaond - Currently, our ArgoCD application paths are all namespace-oriented with differing values files based on environment (development, staging, production, etc). This change breaks with those conventions and would require some other scripting changes to accomodate, so even if this ArgoCD multi-source beta feature is viable (and does not ["change in backwards incompatible ways"](https://argo-cd.readthedocs.io/en/stable/user-guide/multiple_sources/), it's not very desirable. It would be much more operator friendly if there were just a single chart with a convenience flag or two separate charts, one with templated CRDs and one with none.

**Comment by Vinaum8** [comment]
Why not put the option to enable or disable CRD in the Helm Chart?

**Comment by artem-nefedov** [comment]
> Why not put the option to enable or disable CRD in the Helm Chart?

Helm doesn't allow any templating for CRDs that are located in "crds" directory, so it's impossible to affect them based on input values. They can **only** be turned off by `--skip-crds` option (and similar things in various controllers). To make it possible to turn off CRDs via values, they must be moved from "crds" to "templates", like it's done in `karpenter-crd` chart, but that could bring other unwanted side effects (there are reasons why "crds" directory exists and is handled differently).

**Comment by booleanbetrayal** [comment]
> Helm doesn't allow any templating for CRDs that are located in "crds" directory, so it's impossible to affect them based on input values. They can **only** be turned off by `--skip-crds` option (and similar things in various controllers). To make it possible to turn off CRDs via values, they must be moved from "crds" to "templates", like it's done in `karpenter-crd` chart, but that could bring other unwanted side effects (there are reasons why "crds" directory exists and is handled differently).

@artem-nefedov - What are the side effects of templating CRDs? Lots of Helm charts template CRDs, [including ArgoCD itself](https://github.com/argoproj/argo-helm/tree/main/charts/argo-cd/templates/crds).

**Comment by andrewjamesbrown** [comment]
@artem-nefedov unless I'm mistaken, helm supports this via https://github.com/helm/helm/pull/7138 (i.e what ArgoCD is doing)

**Comment by andrewjamesbrown** [comment]
I'm in agreement with @booleanbetrayal here - we use an `ApplicationSet` to setup many applications, and overriding `skipCrds: true` is going to be painful.  It would be far easier to use a similar paradigm like [ArgoCD](https://github.com/argoproj/argo-helm/blob/main/charts/argo-cd/values.yaml#L33) (thanks for the link @booleanbetrayal )

**Comment by artem-nefedov** [comment]
> @artem-nefedov unless I'm mistaken, helm supports this via [helm/helm#7138](https://github.com/helm/helm/pull/7138) (i.e what ArgoCD is doing)

@andrewjamesbrown  This is only for outputting CRDs via `helm template` command. It still doesn't allow any actual templating (changing the content) based on input values.

**Comment by artem-nefedov** [comment]
> @artem-nefedov - What are the side effects of templating CRDs? Lots of Helm charts template CRDs, [including ArgoCD itself](https://github.com/argoproj/argo-helm/tree/main/charts/argo-cd/templates/crds).

@booleanbetrayal For comprehensive reasoning (as the team behind helm sees it), read this:
https://github.com/helm/community/blob/f9e06c16d89ccea1bea77c01a6a96ae3b309f823/architecture/crds.md

**Comment by andrewjamesbrown** [comment]
@artem-nefedov my mistake... ArgoCD CRDs are in the `templates/crds` directory not `crds`
🤦

**Comment by booleanbetrayal** [comment]
> @booleanbetrayal For comprehensive reasoning (as the team behind helm sees it), read this: https://github.com/helm/community/blob/f9e06c16d89ccea1bea77c01a6a96ae3b309f823/architecture/crds.md

I've read this document and I understand some of the theoretical concerns, but when theoretical concerns create practical problems, I think it's reasonable to reassess / challenge the design decisions producing them.

Honestly, I think you all can have the best of both worlds by just omitting the CRDs from the `karpenter` chart and relying on `karpenter-crd` being a dependency.

**Comment by artem-nefedov** [comment]
> @booleanbetrayal For comprehensive reasoning (as the team behind helm sees it), read this: https://github.com/helm/community/blob/f9e06c16d89ccea1bea77c01a6a96ae3b309f823/architecture/crds.md

BTW, I do believe that moving CRDs to "templates" is not a problem for this chart (as long as the team commits to making sure future upgrades don't break anything).
I was just providing some justification on why they could choose otherwise.

**Comment by jmdeal** [comment]
Moving the CRDs to the templates directory like some other projects do could help solve the issues here. This would also enable the convenience flag that @booleanbetrayal wants, because AFAIK items from the values file can only be used to drive templated resources. We're hesitant to make this change because it will break several assumptions:
- If CRDs are already installed in the cluster, helm won't overwrite them when using the main chart
- CRD installation can be skipped using the `--skip-crds` flag

If we were to update the main chart to template CRDs both of those assumptions would be broken. That's not something we want to do lightly.

As for the convenience flag that @booleanbetrayal has brought up before, I believe this would only be possible if the CRDs were defined in our templates folder. If it's possible to drive CRD installation from the `crds` folder from the values file, I don't see a reason why we couldn't do that but I don't think it is. 

We are looking into what options we can pursue here, we understand the limitations of the current helm chart and want to make progress unblocking this.

**Comment by booleanbetrayal** [comment]
Thanks for the consideration behind this @jmdeal !

**Comment by adrianmiron** [comment]
Moving crd's to templates sounds like a plan !

**Comment by artem-nefedov** [comment]
Since this whole conversation started because of hardcoded conversion webhooks, it's worth mentioning that webhook feature is already removed in main and won't cause problems in future releases.

**Comment by adrianmiron** [comment]
> Since this whole conversation started because of hardcoded conversion webhooks, it's worth mentioning that webhook feature is already removed in main and won't cause problems in future releases.

That is good to know, we just need another tag now. 

Would adding the crd chart as a dependency of the main chart now also work ? We would not need to manage two of them and we could template them safely.

**Comment by jmdeal** [comment]
> Moving crd's to templates sounds like a plan!

For the aforementioned reasons, we won't be able to move the CRDs to the templates directory in the main chart. There may be other solutions, but that's not one we can take.

> it's worth mentioning that webhook feature is already removed in main and won't cause problems in future releases.

This is true, but you won't be able to upgrade from a pre-1.0 version of Karpenter to 1.1+ directly. You either need to upgrade through 1.0 to migrate existing resources to the v1 API. Alternatively, you could scale down entirely (remove all v1beta1 NodePools, EC2NodeClasses, and NodeClaims) before upgrading, but that will necessarily incur downtime. 

> Would adding the crd chart as a dependency of the main chart work ?

There's certainly nothing stopping you from doing it yourself, but I don't think we would want to do this for the chart we vend. This would really be the same as just including the crds in the templates directory to begin with, which we can't do for aforementioned reasons.

I don't think there's a realistic path forward for disabling the CRDs through `values.yaml` or removing the static CRDs. Either of these options would require breaking changes to the chart. Discussing with other maintainers, this isn't something we're willing to do given there are quite a few options to work around this issue in Argo. In @booleanbetrayal's case, I think the best option (given you can't use the helm / argo native methods) would be to remove the CRDs folder as part of your ingest process for the Karpenter chart. I understand that this isn't a plug and play solution, but at the end of the day we vend a helm chart, and helm provides native methods to handle this exact case (which Argo natively supports).

**Comment by adrianmiron** [comment]
In the end I created a custom chart from both karpenter charts, set `skip-crds` on the argo app and i can now configure them just fine( the switch from the single chart with default crds went ok ).

A bit off-topic: 

For Karpenter + ArgoCD users, any clue what caused Argo to lose contact with all clusters where the CRDs were updated with the conversion block but the webhook resources themselves were missing ? All resources ( even ones not on karpenter managed nodes ) were showing the webhook CRD lookup error...even though all the resources themselves were ok.  Was the API spammed with the Error ? If so, why did kubectl work fine? Something else ?

**Comment by booleanbetrayal** [comment]
We dug ourselves out of this upgrade mess by just re-packaging the `karpenter` chart minus the `crds/` directory as @jmdeal had suggested for both the `0.37.2` and `1.0.1` releases. The `0.37.2` patch upgrade was seemless, but upon upgrading to `1.0.1` we ran into errors such as the following, which led to the controller not working:

```
{"level":"ERROR","time":"2024-08-29T16:52:58.853Z","logger":"controller","message":"Reconciler error","commit":"62a726c","controller":"nodeclass.hash","controllerGroup":"karpenter.k8s.aws","controllerKind":"EC2NodeClass","EC2NodeClass":{"name":"default"},"namespace":"","name":"default","reconcileID":"c715888a-ded2-4d29-ab81-936fd7be8b77","error":"Internal error occurred: failed calling webhook \"validation.webhook.karpenter.sh\": failed to call webhook: the server rejected our request for an unknown reason; Internal error occurred: failed calling webhook \"validation.webhook.karpenter.sh\": failed to call webhook: the server rejected our request for an unknown reason; Internal error occurred: failed calling webhook \"validation.webhook.karpenter.sh\": failed to call webhook: the server rejected our request for an unknown reason","errorCauses":[{"error":"Internal error occurred: failed calling webhook \"validation.webhook.karpenter.sh\": failed to call webhook: the server rejected our request for an unknown reason"},{"error":"Internal error occurred: failed calling webhook \"validation.webhook.karpenter.sh\": failed to call webhook: the server rejected our request for an unknown reason"},{"error":"Internal error occurred: failed calling webhook \"validation.webhook.karpenter.sh\": failed to call webhook: the server rejected our request for an unknown reason"}]}

{"level":"ERROR","time":"2024-08-29T16:53:00.399Z","logger":"controller","message":"Reconciler error","commit":"62a726c","controller":"nodeclass.status","controllerGroup":"karpenter.k8s.aws","controllerKind":"EC2NodeClass","EC2NodeClass":{"name":"default"},"namespace":"","name":"default","reconcileID":"a92561f9-c494-44b5-b69e-e53edd3e279a","error":"Internal error occurred: failed calling webhook \"defaulting.webhook.karpenter.k8s.aws\": failed to call webhook: the server rejected our request for an unknown reason"}
```

That led to Node scheduling failures and heart palpitations until we stumbled upon [this comment](https://github.com/aws/karpenter-provider-aws/issues/6879#issuecomment-2315685456) by @gnadaban. Upon deleting these validating and mutating webhook resources and restarting the Deployment, things started operating as expected again. 

I couldn't find any sort of reference to this post-install requirement in the [v1 Migration Guide](https://karpenter.sh/docs/upgrading/v1-migration/). The only reference we could find to any sort of manual cleanup of these resources was from this [v0.27.3 upgrade note](https://github.com/aws/karpenter-provider-aws/blob/main/website/content/en/v1.0/upgrading/upgrade-guide.md#upgrading-to-0273). Our first install of Karpenter, having migrated from cluster-autoscaler using [this guide](https://karpenter.sh/docs/getting-started/migrating-from-cas/), was with Karpenter v0.37.0.

I'm leaving this issue open because I believe numerous other people will find themselves in this same mess, and I believe the core concerns still ring true. As it stands, this was the most complicated and frightening upgrade path in my several years of a Kubernetes administrator.

**Comment by gnadaban** [comment]
> That led to Node scheduling failures and heart palpitations until we stumbled upon [this comment](https://github.com/aws/karpenter-provider-aws/issues/6879#issuecomment-2315685456) by @gnadaban.

Wow. I didn’t think this one comment would actually be found by anyone in the sea of issues but I’m glad it was. 

 > As it stands, this was the most complicated and frightening upgrade path in my several years as a Kubernetes administrator.

I had the luck of trying the upgrade on a dev cluster first so I only wrecked and had to recreate the one, but this rings super familiar. I all but gave up and even told my team this is a broken release when I had a surge of defiance and compared the Helm charts of 0.37.2 and 1.0.1 on a whim to find that those webhook resources have been removed, while super misleadingly keeping the webhook.enabled config setting.

Maintainers, when you find this: you should update the migration guide with a glaring red-lettered list which includes this, and the fact that you changed the metrics port. Had to find out the hard way Datadog was no longer collecting metrics, and I don’t like to disappoint when I announce something “successfully completed”. ;)

**Comment by Kavinraja-G** [comment]
This thread sounds like we can't install v1.0 Karpenter in a fresh new cluster too? I'm trying to create few `EC2NodeClass` and it fails on the validation :(

```
➜ kubectl apply -f nodeclass.yaml -n karpenter
Error from server: error when creating "nodeclass.yaml": conversion webhook for karpenter.k8s.aws/v1beta1, Kind=EC2NodeClass failed: Post "https://karpenter.kube-system.svc:8443/conversion/karpenter.k8s.aws?timeout=30s": service "karpenter" not found
```

**Comment by dcherniv** [comment]
Folks, 
You absolutely,100%, never, ever, under any circumstances ship a chart that is not working by itself. 
Omitting CRDs that are required for a chart to function, from the chart is bad.
Suggesting using go and fetch some other chart that needs to be installed as a pre-requisite for your chart is bad.

In other words, you cannot split crds into its own chart and suggest users install a separate chart with crds.
The karpenter chart needs to include CRDs within itself, plain and simple. Just do what argo does and include CRDs in templates/ directory. 

Or look at whatever kube-prometheus-stack chart is doing. Its humongous, yet it works fine more or less and is entirely self-contained. 

Forget about helm's nonsense. Its quite clear that their approach to CRDs has outlived itself. It's likely going to need to be fixed by them anyway.

**Comment by dcherniv** [comment]
> This thread sounds like we can't install v1.0 Karpenter in a fresh new cluster too? I'm trying to create few `EC2NodeClass` and it fails on the validation :(
> 
> ```
> ➜ kubectl apply -f nodeclass.yaml -n karpenter
> Error from server: error when creating "nodeclass.yaml": conversion webhook for karpenter.k8s.aws/v1beta1, Kind=EC2NodeClass failed: Post "https://karpenter.kube-system.svc:8443/conversion/karpenter.k8s.aws?timeout=30s": service "karpenter" not found
> ```

We cannot. Karpenter chart is broken.

**Comment by adrianmiron** [comment]
> This thread sounds like we can't install v1.0 Karpenter in a fresh new cluster too? I'm trying to create few `EC2NodeClass` and it fails on the validation :(
> 
> ```
> ➜ kubectl apply -f nodeclass.yaml -n karpenter
> Error from server: error when creating "nodeclass.yaml": conversion webhook for karpenter.k8s.aws/v1beta1, Kind=EC2NodeClass failed: Post "https://karpenter.kube-system.svc:8443/conversion/karpenter.k8s.aws?timeout=30s": service "karpenter" not found
> ```

You can, but you need to name your karpenter service `karpenter` and install the webhooks from the main chart ( it's disabled by default)

**Comment by povesma** [comment]
> You can, but you need to name your karpenter service `karpenter` and install the webhooks from the main chart ( it's disabled by default)

All what is needed is to add namespace configuration for CRDs in the main chart in the same way as in a separate chart (bec it has default unconfigurable value of `kube-system`, and karpenter is usually installed in NS `karpenter`). Service name configuration would be a bonus.

**Comment by rknightion** [comment]
> and karpenter is usually installed in NS `karpenter`

I thought they changed the recommended namespace to install Karpenter to a few releases ago? As in the usual place to install Karpenter these days is kube-system

**Comment by jonatanzafar59** [comment]
A rule written in blood, never ever upgrade to Karpenter's latest version, always the same story.

**Comment by Vinaum8** [comment]
![image](https://github.com/user-attachments/assets/a6e2d05c-73ba-4341-b9e9-37641670a268)

clearly a problem, even using 2 helm charts.
the helm chart "karpenter": it is not possible to change the service name.
the helm chart "karpenter-crd": it has the correct name, but has a conflict with helm chart karpenter.

The configurations are in conflict.

**Comment by adrianmiron** [comment]
Set `skipcrds = true` of the controller chart

**Comment by dcherniv** [comment]
This is stupid. Why are there crds in the main chart in the first place? 
anyway, we are now firmly in the territory of "implementation details" differences as each system does the render and deploy of charts slightly differently.
on argocd you want `skipCrds` flag to be true on your application spec.
https://argo-cd.readthedocs.io/en/stable/user-guide/application-specification/
I shudder to think how many different CD systems are out there and what their CRDs would have to look like to accomodate this decision by karpenter.

**Comment by Vinaum8** [comment]
In fact, the idea of ​​separating them is not a good one, I may be mistaken due to my lack of knowledge of the product.

However, adding the option to configure the webhook in the main chart is enough and each user can use it as they see fit.

Why this is not done?

**Comment by booleanbetrayal** [comment]
This entire thread and [all](https://github.com/aws/karpenter-provider-aws/issues/6765) [of](https://github.com/aws/karpenter-provider-aws/issues/6867) [the](https://github.com/aws/karpenter-provider-aws/issues/6808) [ancillary](https://github.com/aws/karpenter-provider-aws/issues/6889) [issues](https://github.com/aws/karpenter-provider-aws/issues/6899) that are breaking **real live clusters** could be solved by just making the CRDs in the `karpenter` chart templated, and abandoning the unorthodox 2-chart migration guide (which doesn't work for probably a majority of cluster administrators due to tooling).

Unfortunately, the Karpenter devs are making some sort of weird [ideological last stand](https://github.com/aws/karpenter-provider-aws/issues/6847#issuecomment-2316470591) without regards for the damage they might be causing. I truly believe that given the lack of a real fix for this situation and the litany of problems, someone at AWS should undergo some sort disciplinary review.

**Comment by dcherniv** [comment]
@jmdeal 
> This is true, but you won't be able to upgrade from a pre-1.0 version of Karpenter to 1.1+ directly. You either need to upgrade through 1.0 to migrate existing resources to the v1 API. Alternatively, you could scale down entirely (remove all v1beta1 NodePools, EC2NodeClasses, and NodeClaims) before upgrading, but that will necessarily incur downtime.

Right. On migrating pre-1.0 to version 1.1+, if you go through version 1.0, the webhook wouldn't even be needed, unless i'm misunderstanding something, you can just do the below process?

1. Provision new pools on the new API: v1
2. Keep old pools during transition
3. Cap old pools by setting the limit of current CPU usage so they dont scale up.
4. Cordon old nodes.
5. Bounce the pods at your own pace.
6. Remove all old nodes and pools once all workloads migrate to new node pools.

This upgrade path is very familiar to folks who run kubernetes.
This webhook stuff is just causing folks more pain than just following the normal k8s resource upgrade process.

**Comment by dcherniv** [comment]
I bit the bullet and went with this upgrade process.
So for the folks who are trying to do this on argocd. I had karpenter chart 0.36.1. Here goes my upgrade process:
0. Make sure your IAM role is up to date (there are some new permissions needed)
1. Modify Chart.yaml to bump karpenter to 0.37.3
2. Add karpenter-crd dependency chart on version 0.37.3. Your dependency section should look like this
```
dependencies:
  - name: karpenter-crd
    version: "0.37.3"
    repository: "oci://public.ecr.aws/karpenter"

  - name: karpenter
    version: "0.37.3"
    repository: "oci://public.ecr.aws/karpenter"
```
3. Sync everything with argo. Gawk at the 3 warnings you get. At this point you should see a diff on the CRDs as well before you sync. It should modify your conversion webhook namespace in the CRDs. In my case it switched from `kube-system` to `karpenter`.
![image](https://github.com/user-attachments/assets/e5400fb1-6dfd-4d3f-8c43-343fdc645914)
4. Modify the karpenter argocd application to include `skipCrds: true`. I have an applicationset that is generating my karpenter argocd app. So the manifest should look like this:
![image](https://github.com/user-attachments/assets/e09b94b3-8c36-470f-8664-f78c0fcdce3d)
5. Stare in absolute awe as the warnings should now be gone from your karpenter application in argocd.
6. Check the logs of karpenter pods and test scaling up/down, make sure everything works.
7. Proceed with the upgrade of your stored nodeclasses, pools etc to the new v1 version at your own leisure.

**Comment by dcherniv** [comment]
Had some spare time and decided to press the issue further.
Next steps after the upgrade to 0.37.3:
1. Update your nodepools and ec2nodeclass resources
2. Deploy the updated resources (you should see no diff because the webhook should have taken care of it already)
3. Update the chart for karpenter-crd and karpenter to 1.0.2
4. Make sure to disable the stupid webhook nonsense by passing the following values to your dependent chart:
```
karpenter-crd:
  webhook:
    enabled: false
karpenter:
  webhook:
    enabled: false
```
Now this step is iffy because at first i didn't do this and my karpenter complained that it "failed to validate" my crds for "unknown reason"
I also went ahead and deleted all `mutatingwebhookconfiguration` and `validatingwebhookconfiguration` related to karpenter. This is probably not needed.

I also got some errors due to karpenter not being able to tag an ec2 instance.

**Comment by adrianmiron** [comment]
Is anyone using the community EKS module to spawn the IAM Role + Policy for the controller ? When updating the policy to the v1 version they added a few weeks ago + updating Karpenter to 1.0.2 , it cannot spawn nodes because of lack of permissions ( i did open an issue on their project, just curios )

Also @dcherniv i had a similar experience, but i did find the disabling the webhook from 0.37.X , manually removing the webhooks from the cluster and then updating to 1.0.2 made things smoother ( well, excep[t the inability to spawn new nodes to replace the outdated ones, but that is  not on karpenter i think )

I will come back with updates once i fix the permissions ....

**Comment by booleanbetrayal** [comment]
@dcherniv - We encountered similar permission issues post-upgrade (v1.0.1) when we attempted to migrate our IAM policies from the more lenient [CAS Migration Guide permissions](https://karpenter.sh/preview/getting-started/migrating-from-cas/#create-iam-roles) to the [Cloudformation template](https://raw.githubusercontent.com/aws/karpenter-provider-aws/v1.0.1/website/content/en/preview/getting-started/getting-started-with-karpenter/cloudformation.yaml) values. This is after all NodePools had been migrated to v1, etc. We were forced to revert back to the more permissive policies and are incredibly dubious of anything the Karpenter team instructs us to do at this point.

Here is the (sanitized) permission error we were receiving with the new policies:

```
launching nodeclaim, creating instance, getting launch template configs, getting launch templates, creating launch template, UnauthorizedOperation: You are not authorized to perform this operation. User: arn:aws:sts::123456789012:assumed-role/KarpenterController/1234567890123456789 is not authorized to perform: ec2:CreateLaunchTemplate on resource: arn:aws:ec2:us-east-1:123456789012:launch-template/* because no identity-based policy allows the ec2:CreateLaunchTemplate action.
```

We are assuming it was something about the condition that was failing.

I'd be very curious if you were able to get the Cloudformation bundled policy definitions working and what tweaks you had make, if you pursue that endeavor.

**Comment by dcherniv** [comment]
@booleanbetrayal I can only speculate, as i haven't encountered the error relating to launch-template creation, but briefly looking at the policy here https://raw.githubusercontent.com/aws/karpenter-provider-aws/v1.0.2/website/content/en/preview/getting-started/getting-started-with-karpenter/cloudformation.yaml 
![image](https://github.com/user-attachments/assets/6168d2b7-2026-484d-a53b-c2b3b3ae4957)

I have a sneaking suspicion.
Since the conditions on that policy is via `RequestTag` then i'm pretty sure that karpenter is not passing the required tags in its requests to AWS. (reference: https://docs.aws.amazon.com/IAM/latest/UserGuide/access_tags.html#access_tags_control-requests) 
Or passing tag keys/values that are not explicitly whitelisted in the IAM policy. 
This makes three of us that now have IAM issues :)

**Comment by dcherniv** [comment]
@adrianmiron 
> Is anyone using the community EKS module to spawn the IAM Role + Policy for the controller ? When updating the policy to the v1 version they added a few weeks ago + updating Karpenter to 1.0.2 , it cannot spawn nodes because of lack of permissions ( i did open an issue on their project, just curios )

Yea, i use the https://registry.terraform.io/modules/terraform-aws-modules/eks/aws/latest/submodules/karpenter module.
Upgrade to latest 20.24.0 but there were no meaningful changes to the policies, so i just assumed its out of date or something.

**Comment by dcherniv** [comment]
@adrianmiron 
> @dcherniv The latest version includes an update to the controller IAM role attached policy which is enabled with a variable flag. When i turn that on, the policy gets a long update and it can no longer spawn resources (in either 0.37.X or 1.0.X )
I didn't know this flag existed. Just tested with the flag enabled in the module on 20.24.0, i cannot reproduced. Just scaled up and down one of my node pools and saw launch templates get created and deleted with no problem.

**Comment by dcherniv** [comment]
Some more updates.
I just tried to go from 0.37.3 to 1.0.2 in production and can confidently say there's no valid upgrade path. I tried with webhooks enabled, webhooks disabled. Tried rolling the nodeclaims before going to 1.0.2, tried running 0.37.3 and creating the new nodeclaims with proper v1 spec and migrating to those using phased migration.
There appears to be no way to avoid this error on the nodeclaims
```
{"level":"ERROR","time":"2024-09-11T19:57:52.140Z","logger":"controller","message":"Reconciler error","commit":"b897114","controller":"nodepool.hash","controllerGroup":"karpenter.sh","controllerKind":"NodePool","NodePool":{"name":"default"},"namespace":"","name":"default","reconcileID":"109f1dc7-640d-45ee-85b6-ced5a57d64bb","error":"NodeClaim.karpenter.sh \"default-k8vbn\" is invalid: [spec.nodeClassRef.group: Required value, <nil>: Invalid value: \"null\": some validation rules were not checked because the object was invalid; correct the existing errors to complete validation]; NodeClaim.karpenter.sh \"default-vn6mj\" is invalid: [spec.nodeClassRef.group: Required value, <nil>: Invalid value: \"null\": some validation rules were not checked because the object was invalid; correct the existing errors to complete validation]; NodeClaim.karpenter.sh \"default-2gt6z\" is invalid: [spec.nodeClassRef.group: Required value, <nil>: Invalid value: \"null\": some validation rules were not checked because the object was invalid; correct the existing errors to complete validation]","errorCauses":[{"error":"NodeClaim.karpenter.sh \"default-k8vbn\" is invalid: [spec.nodeClassRef.group: Required value, <nil>: Invalid value: \"null\": some validation rules were not checked because the object was invalid; correct the existing errors to complete validation]"},{"error":"NodeClaim.karpenter.sh \"default-vn6mj\" is invalid: [spec.nodeClassRef.group: Required value, <nil>: Invalid value: \"null\": some validation rules were not checked because the object was invalid; correct the existing errors to complete validation]"},{"error":"NodeClaim.karpenter.sh \"default-2gt6z\" is invalid: [spec.nodeClassRef.group: Required value, <nil>: Invalid value: \"null\": some validation rules were not checked because the object was invalid; correct the existing errors to complete validation]"}]}
```
I dont know what karpenter devs think we are supposed to do, but i'd highly suggest to not upgrade to 1.x.x, it is an absolute disaster.

**Comment by dcherniv** [comment]
> They expect us to [re-package the Chart archives](https://github.com/aws/karpenter-provider-aws/issues/6847#issuecomment-2316470591) (and then subsequently ignore all checksum validation during `helm dep update` calls).

This is not to mention the problem with IAM policy. I reproduced it. 1.0.2 policy is not backwards compatible with 0.37.3 Karpenter. As soon as I enabled that v1 flag in my IAM module while still keeping Karpenter itself on 0.37.3 it promptly broke with that error lacking permissions for launch templates and ec2 tagging. 
What a clown show this upgrade is

**Comment by bryantbiggs** [comment]
let keep the convo civil and professional please - we should assume good intent. I can vouch for the Karpenter team in terms of how much time and effort went into not only creating the v1.0 release, but testing and validating it.

Please continue to share any findings or steps taken to resolve as this will help both the Karpenter team and other users who may run into similar challenges

**Comment by dcherniv** [comment]
@bryantbiggs 
> let keep the convo civil and professional please - we should assume good intent. I can vouch for the Karpenter team in terms of how much time and effort went into not only creating the v1.0 release, but testing and validating it.

I apologize, but you have to get in my shoud. This is extremely frustrating for an end user, going through upgrade path https://karpenter.sh/v1.0/upgrading/v1-migration/#upgrade-procedure, it just doesn't work.
I'm happy to share exactly when and how it is breaking, but it is the happy path, and even that doesn't work.

**Comment by booleanbetrayal** [comment]
@bryantbiggs - Do you have authority in the Karpenter release process? Can you ensure that ArgoCD and Flux both become test vectors for future releases?

**Comment by dcherniv** [comment]
@bryantbiggs 
Created additional specific issue i'm having currently: https://github.com/aws/karpenter-provider-aws/issues/6981
If we can resolve this, then at least there's a semblance of a path forward. Right now it's just filling up the logs with errors.

**Comment by dcherniv** [comment]
I finally resolved my issues in prod with stale nodeclaims RE: https://github.com/aws/karpenter-provider-aws/issues/6981

This was by strategically terminating nodes in aws console whose nodeclaim didn't have `group` field for some reason:
```
for i in `kubectl get nodeclaims | grep -v NAME | awk '{print $1}'` ; do echo $i ; kubectl get nodeclaim -o yaml $i | grep group ; done
default-arm64-6pv26
    group: karpenter.k8s.aws
default-arm64-8xdhn
    group: karpenter.k8s.aws
default-arm64-9xmjx
    group: karpenter.k8s.aws
default-arm64-fcdc7
    group: karpenter.k8s.aws
default-arm64-kndsr
    group: karpenter.k8s.aws
default-arm64-vrzlw
    group: karpenter.k8s.aws
default-bvqlp
    group: karpenter.k8s.aws
default-h74m4
[...]
for the nodeclaims that didn't have group set in the above output, i had to terminate the instance forcefully through aws console.
Once karpenter provisioned replacements for these, i quickly reverted back to 0.37.3, with webhooks enabled (because 0.37.3 with webhooks disabled doesn't even work (see this one https://github.com/aws/karpenter-provider-aws/issues/6975) 
0.37.3 cleaned up stale claims with no problems. Carefully monitoring the nodeclaims (You dont want 0.37.3 to provision new claims because 1.0.2 wont be able to delete them so you'd be at square 1). Anyway, you gotta be quick here.
I executed the "procedure" real quick:
```
  kubectl delete validatingwebhookconfiguration validation.webhook.config.karpenter.sh
  kubectl delete validatingwebhookconfiguration validation.webhook.karpenter.k8s.aws
  kubectl delete validatingwebhookconfiguration validation.webhook.karpenter.sh
  kubectl delete mutatingwebhookconfiguration defaulting.webhook.karpenter.k8s.aws
```
Then quickly synced to 1.0.2 in argocd with prune option.
Finally, i have nodeclaims on the new v1 spec and no errors in karpenter logs.
And don't forget to sync the IAM policy to v1 version because v1 policy is not compatible with 0.37.3 (see https://github.com/aws/karpenter-provider-aws/issues/6983)

**Comment by jonathan-innis** [comment]
For a lot of folks here, it seems like the validating and mutating webhooks are hanging around even after the upgrade to v1.0 where they are dropped in the chart. Is this expected behavior for Argo? I'd expect that when a resource is removed from the helm release that the respective resource is also pruned by whichever GitOps tool is applying it.

**Comment by booleanbetrayal** [comment]
@jonathan-innis - ArgoCD has an optional [Prune](https://argo-cd.readthedocs.io/en/stable/user-guide/sync-kubectl/#prune) option in syncing, which will delete any [tracked resources](https://argo-cd.readthedocs.io/en/latest/user-guide/resource_tracking/) that are presently missing from the chart manifests (due to templating or value conditions). Until they are Pruned, they will reflect in the ArgoCD UI as a diff with missing contents. However, in our case, **I  do not believe we had any visibility in our Application that any `ValidatingWebhookConfigurations` or `MutatingWebhookConfigurations` ever existed.**

If some other process had updated the `metadata.labels` values on those resources, or [they had been explicitly overwritten by the chart manifest themselves](https://github.com/aws/karpenter-provider-aws/blob/v0.37.1/charts/karpenter/templates/_helpers.tpl#L33-L46) _(possibly?)_ then ArgoCD would lose the ability to track them as they would become detached from ArgoCD management. Therefore would not show up in the Application at all, let alone any indication of diff, and would not cleaned up when Prune was enabled.

I am theorizing here that there was some label mismanagement as part of the numerous charts / versions / hooks / templating helpers that broke these resources at deployment onset, but I'm definitely not rewinding the clock on attempting to reproduce this one. 

I think it really behooves the Karpenter team to incorporate ArgoCD and Flux tooling into release testing procedures (cc: @bryantbiggs)

**Comment by adrianmiron** [comment]
> I finally resolved my issues in prod with stale nodeclaims RE: #6981
> 
> Issue #6981 is likely caused, in my case, by the fact that conversion webhooks do NOT work on karpenter 1.0.2 chart: #6982 If you manage to get conversion webhook working on 1.0.2 then the below probably wouldn't apply to you.
> 
> Warning: below step may incur downtime if you go too fast! As well i wouldn't advise to go this route if you have a large cluster. Luckily mine was a handful of nodes, which still took me a better part of the day. This was by strategically terminating nodes in aws console whose nodeclaim didn't have `group` field for some reason:
> 
> ```
> for i in `kubectl get nodeclaims | grep -v NAME | awk '{print $1}'` ; do echo $i ; kubectl get nodeclaim -o yaml $i | grep group ; done
> default-arm64-6pv26
>     group: karpenter.k8s.aws
> default-arm64-8xdhn
>     group: karpenter.k8s.aws
> default-arm64-9xmjx
>     group: karpenter.k8s.aws
> default-arm64-fcdc7
>     group: karpenter.k8s.aws
> default-arm64-kndsr
>     group: karpenter.k8s.aws
> default-arm64-vrzlw
>     group: karpenter.k8s.aws
> default-bvqlp
>     group: karpenter.k8s.aws
> default-h74m4
> [...]
> ```
> 
> for the nodeclaims that didn't have group set in the above output, i had to terminate the instance forcefully through aws console. Once karpenter provisioned replacements for these, i quickly reverted back to 0.37.3, with webhooks enabled (because 0.37.3 with webhooks disabled doesn't even work (see this one #6975) 0.37.3 cleaned up stale claims with no problems. Carefully monitoring the nodeclaims (You dont want 0.37.3 to provision new claims because 1.0.2 wont be able to delete them so you'd be at square 1). Anyway, you gotta be quick here. I executed the "procedure" real quick:
> 
> ```
>   kubectl delete validatingwebhookconfiguration validation.webhook.config.karpenter.sh
>   kubectl delete validatingwebhookconfiguration validation.webhook.karpenter.k8s.aws
>   kubectl delete validatingwebhookconfiguration validation.webhook.karpenter.sh
>   kubectl delete mutatingwebhookconfiguration defaulting.webhook.karpenter.k8s.aws
> ```
> 
> Then quickly synced to 1.0.2 in argocd with prune option. Finally, i have nodeclaims on the new v1 spec and no errors in karpenter logs. And don't forget to sync the IAM policy to v1 version because v1 policy is not compatible with 0.37.3 (see #6983)
> 
> To go back and forth between 0.37.3 and 1.0.2 i suggest removing ALL of the conditionals manually in the IAM policy for the time. This is outline in step 8 here: https://karpenter.sh/v1.0/upgrading/v1-migration/#upgrade-procedure Unfortunately it fails to mention that policy is completely incompatible. 0.37.3 doesn't work with policy for 1.0.2 and vice-versa
> 
> PS. Don't forget to disable the webhooks on version 1.0.2 when you are going back and forth. I suggest having two values files prepared for this case (see #6982)
> 
> Good luck!

When doing rollback to remoe stale NodeClaims, do it to 0.37.2 , with webhooks disabled, that worked for me.

 However what happens if during the switch your cluster wants to spawn nodes...you get 0.37.2 NodeClaims again :)) ..... unless you setup different NodePools for both versions, block 0.37 NodePools from creating nodes when moving to 1.X and keep that setting when going back to 0.37 to avoid spawing bad nodes...in other words it is not worth doing on 10 clusters

**Comment by dcherniv** [comment]
> When doing rollback to remoe stale NodeClaims, do it to 0.37.2 , with webhooks disabled, that worked for me.

0.37.2 worked. I just tried following the v1 upgrade guide and there it was mentioned that you have to be on the latest patch version, which was unfortunately broken :(

**Comment by dcherniv** [comment]
I think this has something to do with https://kubernetes.io/docs/concepts/architecture/garbage-collection/#orphaned-dependents maybe?
since the webhook specifies karpenter ?deployment/rs? as its owner in the owner block:
```
  ownerReferences:
  - apiVersion: v1
    blockOwnerDeletion: true
    controller: true
    kind: Namespace
    name: karpenter
    uid: a4d1f1ad-a593-4666-8677-62337f9c602f
```
But when we flip webhook.enabled flag in the chart values this causes the re-creation of the karpenter replicaset (since it goes by uid, its hard to tell who is the owner, whether its the rs or the deployment)
When the rs is deleted the webhooks become orphans and are removed from consideration or something.
i'm not 100% on how this owner reference block works, but it seems plausible.

**Comment by jonathan-innis** [comment]
> I am theorizing here that there was some label mismanagement as part of the numerous charts / versions / hooks / templating helpers that broke these resources at deployment onset, but I'm definitely not rewinding the clock on attempting to reproduce this one.

That would be my guess as well. We need to look into the reconciler in the code that's injecting the certificate into the Validatin and Mutating Webhook Configurations. I suspect that it may be doing exactly what you are saying and removing ownership. And solving that problem would reduce a lot of the pain here.

> I think it really behooves the Karpenter team to incorporate ArgoCD and Flux tooling into release testing procedures

I'd be interested to see what other projects do around testing with GitOps tooling integrations. I'd imagine that there's some prior art out there for integration testing but I'd be interested to see which CNCF projects have it.

**Comment by iNoahNothing** [comment]
FWIW I found an easier way to recover from this
1. `kubectl delete` the `NodeClaim`s that were left over from earlier karpenter deploy
2. Temporarily update the `nodeclaims.karpenter.sh` crd to make `group` not a required field so that you can edit the crd without failing CEL validation
3. `kubectl edit` the `NodeClaim`s and remove the finalizer so that the resource gets removed from the API server
4. Restart karpenter and let it spin up new nodes

This will temporarily orphan the old nodes that were claimed by the deleted `NodeClaim`s but Karpenter appears to cleanly remove those for me once it was back in a healthy state. This was relatively painless and did not incur downtime

**Comment by adrianmiron** [comment]
> FWIW I found an easier way to recover from this
> 
> 1. `kubectl delete` the `NodeClaim`s that were left over from earlier karpenter deploy
> 2. Temporarily update the `nodeclaims.karpenter.sh` crd to make `group` not a required field so that you can edit the crd without failing CEL validation
> 3. `kubectl edit` the `NodeClaim`s and remove the finalizer so that the resource gets removed from the API server
> 4. Restart karpenter and let it spin up new nodes
> 
> This will temporarily orphan the old nodes that were claimed by the deleted `NodeClaim`s but Karpenter appears to cleanly remove those for me once it was back in a healthy state. This was relatively painless and did not incur downtime

That sounds clever, but i would  **not** do that on my prod clusters :D

**Comment by iNoahNothing** [comment]
> > FWIW I found an easier way to recover from this
> > 
> > 1. `kubectl delete` the `NodeClaim`s that were left over from earlier karpenter deploy
> > 2. Temporarily update the `nodeclaims.karpenter.sh` crd to make `group` not a required field so that you can edit the crd without failing CEL validation
> > 3. `kubectl edit` the `NodeClaim`s and remove the finalizer so that the resource gets removed from the API server
> > 4. Restart karpenter and let it spin up new nodes
> > 
> > This will temporarily orphan the old nodes that were claimed by the deleted `NodeClaim`s but Karpenter appears to cleanly remove those for me once it was back in a healthy state. This was relatively painless and did not incur downtime
> 
> That sounds clever, but i would **not** do that on my prod clusters :D

Curious as to why this would be thought of as riskier to you than force deleting the nodes and having to the delicate and well timed dance of rolling back and rolling forward described above. Since the nodes will be drained before being removed everything was restarted and running before the nodes were reaped.

**Comment by Vinaum8** [comment]
> Well i would not use either solution on prod, i will wait for an official version which does this for us without any tricks. We do not **need** to move to 1.0.X right now and most of the stuff here is just to discover issues and possible solutions so the devs know about them

me too

**Comment by jonathan-innis** [comment]
> That would be my guess as well. We need to look into the reconciler in the code that's injecting the certificate into the Validation and Mutating Webhook Configurations. I suspect that it may be doing exactly what you are saying and removing ownership. And solving that problem would reduce a lot of the pain here.

So at least for this specific issue, I think we've tracked down what's going on here.

There's an open issue [here](https://github.com/knative/serving/issues/15483) that actually specifically describes this interaction between Knative and ArgoCD. What it comes down to is that ArgoCD refuses to delete objects with ownerReferences -- even if it was the original creator of the object.

Because knative injects an owner reference onto the `ValidatingWebhookConfiguration` and `MutatingWebhookConfiguration` as part of its reconciliation flow, this causes ArgoCD to ignore these objects when pruning and leave them behind on the cluster. I'm trying out what happens when we don't have these owner references by patching the knative package that we are using here: https://github.com/aws/karpenter-provider-aws/pull/7042 to see if that resolves the issue (I suspect it should).

Code link to where ArgoCD does this check: https://github.com/argoproj/gitops-engine/blob/master/pkg/cache/cluster.go#L1167

**Comment by jonathan-innis** [comment]
Trying to get this feature into upstream `knative/pkg` so we can add the change and patch this back to the Karpenter versions that are affected by this: https://github.com/knative/pkg/pull/3095

**Comment by ShahroZafar** [comment]
Hi, we are planning to upgrade to v1 but before upgrading I landed on this issue.

We are using ArgoCD for managing karpenter controller. We have 2 applications
1. Karpenter controller 
2. Nodepools and ec2nodeclasses

We plan to go from v0.37.3 (We upgraded to this version just because it was mentioned in the upgrade guide. Had to manually update the namespace to karpenter in the CRD to fix the karpenter service not found issue in the CRDs ) to v1.0.2. 

I see that conversion webhooks are removed in v1.0.2 helm chart. Since the webhooks are removed (Though this issue points correctly that they are just removed from the argocd when you put webhook enabled false and not actually which i confirmed on v0.37.3 as well), how the webhook conversion is supposed to work.

I do see webhooks in our argocd application in v0.37.3.

**Comment by jonathan-innis** [comment]
We just released latest patch versions of pre-v1.0 versions that fix this issue so that these configuration resources aren't leaked. Please use one of the following versions prior to going to 1.0.x since these versions remove the ownerReference that is causing Argo to leak the resources and causes the failure on upgrade:

v0.33.10
v0.34.11
v0.35.10
v0.36.7
v0.37.5

**Comment by laserpedro** [comment]
hello, a little summary that may help here as a feedback for our v1 upgrade:

env:
- EKS 1.30
- Karpenter deployed via ArgoCD / Karpenter node classes and nodepools deployed also with ArgoCD.

Feedback:
upgrade to minor patch 0.37.5 to 1.0.6 enables a clean deletion of the webhooks of karpenter : validating and mutating.
Though we are encountering a failure of the migration webhook to v1 of the CRDs: the only error we can see from the controller's log regarding the webhook is detailed here: https://github.com/aws/karpenter-provider-aws/issues/6898 , but eventually our CRDs are still all in v1beta1. 
And also IDK if this is an issue specific to our network policies (explicit deny on the creation of public IP address), but with the latest v1 policy (the one in the migration guide or the one in the eks tf module, submodule /karpenter) we receive a permission denied error now whereas 0.37.X [policy](https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/modules/karpenter/policy.tf#L5) was not generating this error:
`You are not authorized to perform this operation. User: arn:aws:sts::XXXX:assumed-role/KarpenterController-20240626054425729500000022/1728281406041809636 is not authorized to perform: ec2:RunInstances on resource: arn:aws:ec2:me-south-1:XXXX:network-interface/* with an explicit deny in a service control policy` (subnets have no auto assignment of public IPV4 address ....): so here there is sth strange to me.

**Comment by ravigude** [comment]
Any one upgraded from v0.37.5 to 1.0.2 ?

**Comment by adrianmiron** [comment]
> Any one upgraded from v0.37.5 to 1.0.2 ?

Yes, it worked for me like this : 

- upgrade to 0.37.5 ( with webhooks enabled , with them off you get a RBAC error...needs a fix )
- upgrade resources + IAM policy to v1.0.X version
- upgrade to v1.0.6 ( do not disable the webhook in de CRD yet, it will trigger a bunch of errors)
- disable the webhook inside the CRD

**Comment by jmdeal** [comment]
Hi @laserpedro, I followed up in #6898, I don't believe what you are seeing indicates an issue in the conversion webhook. As for the network policy issue, this is unrelated to Argo. This was a change made at v1 to reduce implicit field management since `associatePublicIPAddress` is now available on the `EC2NodeClass` (#6213). This could be a breaking change for users with SCPs and should be called out in the upgrade guide, I'll cut a docs PR to get it in.

**Comment by Vinaum8** [comment]
I decided work with two helm charts in my argocd server:

- karpenter/karpenter
- karpenter/karpenter-crd

With configurations in the applicationsets:

```
ignoreDifferences:
  - group: apiextensions.k8s.io
    kind: CustomResourceDefinition
    name: ec2nodeclasses.karpenter.k8s.aws
    jqPathExpressions:
    - .spec.conversion.webhook.clientConfig
  - group: apiextensions.k8s.io
    kind: CustomResourceDefinition
    name: nodeclaims.karpenter.sh
    jqPathExpressions:
    - .spec.conversion.webhook.clientConfig
  - group: apiextensions.k8s.io
    kind: CustomResourceDefinition
    name: nodepools.karpenter.sh
    jqPathExpressions:
    - .spec.conversion.webhook.clientConfig
 ```

**Karpenter-crd:**
```
webhook:
    # -- Whether to enable the webhooks.
    enabled: true
    serviceName: insert_here_correct_name_service_karpenter
    serviceNamespace: insert_here_correc_name_namespace_karpenter
    # -- The container port to use for the webhook.
    port: 8443
```

**Comment by icicimov** [comment]
What a mess ... BTW why do we have to have the finalizer:
```
  finalizers:
    - karpenter.k8s.aws/termination
 ```
on the EC2NodeClass? If one wants to move from `role` to `instanceProfile` or vice-versa we need to delete it first which is not possible without deleting the `ec2nodeclasses.karpenter.k8s.aws` CRD first which is kinda convoluted?

**Comment by tvandinther** [comment]
Is there any way in which to use the helm chart and Argo CD to make the upgrade to v1 if you have karpenter deployed to a namespace *other* than `kube-system`? The conversion webhook configuration in the CRDs specify `kube-system` as the namespace and the upgrade guide offers a step to patch this to another namespace but this is not feasible in a GitOps workflow. From my understanding, this upgrade is not possible with GitOps for as long as the conversion webhooks are part of the helm chart. We need to be able to disable these with a helm value so that we can make a commit with the resource conversion ourselves.

**Comment by burner1024** [comment]
Even if not GitOps, at least please provide a clear, official upgrade path which takes into account all the quirks mentioned here.
Based on how long this has been around, I'm starting to wonder if Karpenter is the right choice at all.

**Comment by trojaond** [comment]
I can report a successful and seamless upgrade from version `0.37.2` ->`0.37.6` and then to -> `1.0.8`, without any manipulation of webhooks, CRDs, or ignored fields.

I am deploying Karpenter as 2 separate ArgoCD applications:

- Karpenter - `karpenter/karpenter` set  `skipCrds: true` to avoid installing CRDs  
- Karpenter CRD - `karpenter/karpenter-crd`

Here are the steps that I followed:

1. Splitting Karpenter into 2 application
2. Updated the Karpenter CRD to the latest pre-1.x version, `0.37.6`.
3. Upgraded Karpenter itself to version `0.37.6`, matching the CRD.
4  Update the Karpenter CRD from version `0.37.6` to `1.0.8`.
5. Finally update Karpenter from version `0.37.6` to `1.0.8`.

**Comment by Vinaum8** [comment]
> I can report a successful and seamless upgrade from version `0.37.2` ->`0.37.6` and then to -> `1.0.8`, without any manipulation of webhooks, CRDs, or ignored fields.
> 
> I am deploying Karpenter as 2 separate ArgoCD applications:
> 
>     * Karpenter - `karpenter/karpenter` set  `skipCrds: true` to avoid installing CRDs
> 
>     * Karpenter CRD - `karpenter/karpenter-crd`
> 
> 
> Here are the steps that I followed:
> 
>     1. Splitting Karpenter into 2 ArgoCD applications
> 
>     2. Updated the Karpenter CRD to the latest pre-1.x version, `0.37.6`.
> 
>     3. Upgraded Karpenter itself to version `0.37.6`, matching the CRD.
>        4  Update the Karpenter CRD from version `0.37.6` to `1.0.8`.
> 
>     4. Finally update Karpenter from version `0.37.6` to `1.0.8`.

I did this too, I migrated from 1 application to 2 applications.

Karpenter and karpenter-crd

**Comment by petewilcock** [comment]
Really confused here as to why we're not just adding a value in the main karpenter chart for `installCrds: false` (rather than rely on Helm functionality) which then simply skips the Crd installation on the main chart. I expect it to be such a common use case to want Karpenter as a single application which includes CRDs and main chart together. Using the global helm argument to skip all CRD installations clearly doesn't work here and even GitOps tools can't slice in this distinction between indivudual dependencies. 

It is either this or take the stance that the main chart does not install CRDs at all, and must be installed from a separate chart first as a dependency.

**Comment by rschalo** [comment]
Provided the workarounds given in this thread - marking as solved. Please feel free to join working group to discuss if this is a hard blocker for upgrades.
